pkg/storage/stores/shipper/uploads/table.go

```diff
@@ -0,0 +1,321 @@
+		path:              path,
+		uploader:          uploader,
+		storageClient:     storageClient,
+		boltdbIndexClient: boltdbIndexClient,
+		dbs:               dbs,
+		uploadedDBsMtime:  map[string]time.Time{},
+	}, nil
+}
+
+// Query serves the index by querying all the open dbs.
+func (lt *Table) Query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	lt.dbsMtx.RLock()
+	defer lt.dbsMtx.RUnlock()
+
+	for _, db := range lt.dbs {
+		if err := lt.boltdbIndexClient.QueryDB(ctx, db, query, callback); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (lt *Table) getOrAddDB(name string) (*bbolt.DB, error) {
+	lt.dbsMtx.Lock()
+	defer lt.dbsMtx.Unlock()
+
+	var (
+		db  *bbolt.DB
+		err error
+		ok  bool
+	)
+
+	db, ok = lt.dbs[name]
+	if !ok {
+		db, err = local.OpenBoltdbFile(filepath.Join(lt.path, name))
+		if err != nil {
+			return nil, err
+		}
+
+		lt.dbs[name] = db
+		return db, nil
+	}
+
+	return db, nil
+}
+
+// Write writes to a db locally with write time set to now.
+func (lt *Table) Write(ctx context.Context, writes local.TableWrites) error {
+	return lt.write(ctx, time.Now(), writes)
+}
+
+// write writes to a db locally. It shards the db files by truncating the passed time by shardDBsByDuration using https://golang.org/pkg/time/#Time.Truncate
+// db files are named after the time shard i.e epoch of the truncated time.
+// If a db file does not exist for a shard it gets created.
+func (lt *Table) write(ctx context.Context, tm time.Time, writes local.TableWrites) error {
+	shard := fmt.Sprint(tm.Truncate(shardDBsByDuration).Unix())
+
+	db, err := lt.getOrAddDB(shard)
+	if err != nil {
+		return err
+	}
+
+	return lt.boltdbIndexClient.WriteToDB(ctx, db, writes)
+}
+
+// Stop closes all the open dbs.
+func (lt *Table) Stop() {
+	lt.dbsMtx.Lock()
+	defer lt.dbsMtx.Unlock()
+
+	for name, db := range lt.dbs {
+		if err := db.Close(); err != nil {
+			level.Error(util.Logger).Log("msg", fmt.Errorf("failed to close file %s for table %s", name, lt.name))
+		}
+	}
+
+	lt.dbs = map[string]*bbolt.DB{}
+}
+
+// RemoveDB closes the db and removes the file locally.
+func (lt *Table) RemoveDB(name string) error {
+	lt.dbsMtx.Lock()
+	defer lt.dbsMtx.Unlock()
+
+	db, ok := lt.dbs[name]
+	if !ok {
+		return nil
+	}
+
+	err := db.Close()
+	if err != nil {
+		return err
+	}
+
+	delete(lt.dbs, name)
+
+	return os.Remove(filepath.Join(lt.path, name))
+}
+
+// Upload uploads all the dbs which are never uploaded or have been modified since the last batch was uploaded.
+func (lt *Table) Upload(ctx context.Context) error {
+	lt.dbsMtx.RLock()
+	defer lt.dbsMtx.RUnlock()
+
+	for name, db := range lt.dbs {
+		stat, err := os.Stat(db.Path())
+		if err != nil {
+			return err
+		}
+
+		lt.uploadedDBsMtimeMtx.RLock()
+		uploadedDBMtime, ok := lt.uploadedDBsMtime[name]
+		lt.uploadedDBsMtimeMtx.RUnlock()
+
+		if ok && !uploadedDBMtime.Before(stat.ModTime()) {
+			continue
+		}
+
+		err = lt.uploadDB(ctx, name, db)
```

Comment:
Same I think we should upload multiple DB at once. Not easy to write so we can start with this if you prefer. But you're less at risk of corrupted data if you run faster :)

Reply:
Same as above, let us do the optimizations when we need them because they add complexity and would also, in turn, increase the size of the PR.

Reply:
Sure !